src/4.0/wrap.ts

```diff
@@ -0,0 +1,105 @@
+import { hashToBuffer, isStringArray, SchemaValidationError } from "../shared/utils";
+import { MerkleTree } from "../shared/merkle";
+import { ContextUrl, SchemaId } from "../shared/@types/document";
+import { WrappedDocument } from "./types";
+import { digestCredential } from "../4.0/digest";
+import { validateSchema as validate } from "../shared/validate";
+import { WrapDocumentOptionV4 } from "../shared/@types/wrap";
+import { OpenAttestationDocument } from "../__generated__/schema.4.0";
+import { encodeSalt, salt } from "./salt";
+import { validateW3C } from "./validate";
+import { getSchema } from "../shared/ajv";
+
+export const wrapDocument = async <T extends OpenAttestationDocument>(
+  credential: T,
+  options: WrapDocumentOptionV4 // eslint-disable-line @typescript-eslint/no-unused-vars
+): Promise<WrappedDocument<T>> => {
+  const document = { ...credential };
+
+  // 1. Ensure that required @contexts are present and in the correct order
+  // @context: [Base, OA, ...]
+  const contexts = new Set(["https://www.w3.org/2018/credentials/v1", ContextUrl.v4_alpha]);
+
+  if (typeof document["@context"] === "string") {
+    contexts.add(document["@context"]);
+  } else if (isStringArray(document["@context"])) {
+    document["@context"].forEach((context) => contexts.add(context));
+  }
+
+  // Since JavaScript Sets preserve insertion order and duplicated inserts do not affect the order, we can do this:
+  document["@context"] = Array.from(contexts);
+
+  // 2. Ensure that required types are present and in the correct order
+  // type: ["VerifiableCredential", "OpenAttestationCredential", ...]
+  const types = new Set(["VerifiableCredential", "OpenAttestationCredential"]);
+
+  if (typeof document["type"] === "string") {
+    types.add(document["type"]);
+  } else if (isStringArray(document["type"])) {
+    document["type"].forEach((type) => types.add(type));
+  }
+
+  // Since JavaScript Sets preserve insertion order and duplicated inserts do not affect the order, we can do this:
+  document["type"] = Array.from(types);
```

Comment:
Assuming if we use the above helper function, we can turn this part into just this:
```
document["type"] = mergeDocSet(document["type"], ["VerifiableCredential", "OpenAttestationCredential"]);
```